### PR TITLE
Update to use forwardRef on withTheme, and ref related typings

### PR DESCRIPTION
**Pull request by nathanmarks**

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ import spacing from './styles/spacing'
 import PhoneNumberField from './components/Forms/PhoneNumberField'
 import DateField from './components/Forms/DateField'
 import MaskedTextField from './components/Forms/MaskedTextField'
-import TextField from './components/Forms/TextField'
+import TextField, { TextFieldRefApi } from './components/Forms/TextField'
```
Comment (mmarcuccio) [inline src/index.d.ts]: Was `TextFieldRefApi` removed?

Comment (nathanmarks) [reply src/index.d.ts]: yep, thanks!

**Comment by mmarcuccio** [comment]
Should we change https://github.com/instacart/Snacks/blob/0be96fad6ead7fbb52df93b2cf83f822a9c8e4de/src/components/Forms/FormComponent.d.ts#L39 ?

File: src/components/Forms/TextField.d.ts [hunk 2]
```diff
@@ -69,12 +69,21 @@ export interface TextFieldProps
 
   /** Any additonal props to add to the element (e.g. data attributes). */
   elementAttributes?: ElementAttributes<React.ComponentPropsWithoutRef<'input'>>
-}
 
-declare class TextFieldBase extends React.Component<TextFieldProps> {
-  triggerFocus(): void
+  // ref?: React.Ref<{
```
Comment (mmarcuccio) [inline src/components/Forms/TextField.d.ts]: Should probably remove this extra comment.

File: src/components/Forms/FormComponent.d.ts [hunk 3]
```diff
@@ -36,40 +36,42 @@ interface FormComponentProps {
...
+
+type FormComponentRef<P> = {
+  ref?: React.Ref<{
+    FormComponent: 'ref' extends keyof P
+      ? P extends { ref?: React.Ref<infer R> }
+        ? R
+        : never
+      : never
+  }>
+}
```
Comment (mmarcuccio) [inline src/components/Forms/FormComponent.d.ts]: Can you explain this one a bit?

File: src/components/Forms/FormComponent.d.ts [hunk 4]
```diff
@@ -41,23 +42,19 @@ export interface FormComponentInjectedProps extends FormComponentProps {
...
 
-type FormComponentRef<P> = {
+type FormComponentRef<T extends React.ComponentType<any>> = {
   ref?: React.Ref<{
-    FormComponent: 'ref' extends keyof P
-      ? P extends { ref?: React.Ref<infer R> }
-        ? R
-        : never
-      : never
+    FormComponent: GetRef<T>
```
Comment (mmarcuccio) [inline src/components/Forms/FormComponent.d.ts]: Much easier to understand, thanks!

**Pull request merged**